src/librustc/traits/fulfill.rs

```diff
@@ -99,13 +104,98 @@ pub struct PendingPredicateObligation<'tcx> {
     pub stalled_on: Vec<Ty<'tcx>>,
 }
 
+/// An obligation which cannot be fulfilled in the context
+/// it was registered in, such as auto trait obligations on
+/// `impl Trait`, which require the concrete type to be
+/// available, only guaranteed after finishing type-checking.
+#[derive(Clone, Debug)]
+pub struct DeferredObligation<'tcx> {
+    pub predicate: ty::PolyTraitPredicate<'tcx>,
+    pub cause: ObligationCause<'tcx>
+}
```

Comment:
AFAIK, even `<() as Foo>::X = u32` won't type-check at this moment, so such speculations don't even come into play at this moment - although something like the lattice rule may add them.